### PR TITLE
Fix validParams brain damage 

### LINKED ISSUE #10586: Fix validParams Declaration Errors
## Rationale
Declarations of validParams are required in the header files.  However, we've been getting away with omitting it in many circumstances.  That needs to be fixed!

## Description
We should make it a compile error if you forget or mess up the declaration of your valid params.  We'll do this by adding a `static_assert` to the generic definition of validParams()

## Impact
Reduce error rate.

### PR DESCRIPTION
**Pull request by friedmud**
Fix all of the validParams errors in MOOSE, test and Modules.  Also: change the default implementation of `validParams()` to disallow this in the future.

fixes #10586

### CODE REVIEW

File: framework/src/actions/MooseObjectAction.C [hunk 1]
```diff
@@ -27,6 +28,9 @@ validParams<MooseObjectAction>()
   return params;
 }
 
+template <>
+InputParameters validParams<MooseObject>();
```
Comment (andrsd) [inline framework/src/actions/MooseObjectAction.C]:
```
template <>
InputParameters validParams<MooseObjectAction>()
{
  InputParameters params = validParams<Action>();
  return params;
}
```
?

File: framework/include/utils/InputParameters.h [hunk 2]
```diff
@@ -810,11 +810,7 @@ class InputParameters : public Parameters
   bool _allow_copy;
 
   // These are the only objects allowed to _create_ InputParameters
-  friend InputParameters validParams<MooseObject>();
-  friend InputParameters validParams<Action>();
-  friend InputParameters validParams<Problem>();
```
Comment (andrsd) [inline framework/include/utils/InputParameters.h]: 👍

**Review by andrsd — DISMISSED** [review]

**Comment by andrsd** [comment]
So, with this change, even classes that were not registered (like intermediate classes that provide interface, but are not meant for end usage) will have to provide `validParams`?

**Comment by friedmud** [comment]
@andrsd They already do.  Any class that inherits from MooseObject must provide validParams.  An intermediate base will have it's validParams function called by the classes that inherit from it.

**Comment by gambka** [comment]
While I understand the requirement of including the declaration of the validParams() in the header file but I was wondering if you could provide clarification on the following.  For example in the file you linked in the Moose Public Service Announcement the validParams function is not omitted but rather declared at the bottom of the header file. If I look at the file in this pull request the following appears on lines 16 and 17 as well as 38 and 39 of CompositeEigenstrain.h

```
 template <>
 InputParameters validParams<CompositeEigenstrain>();
```

My understanding was that if declared at the end of the header file that the forward declaration of `class CompositeEigenstrain;` would not be required.  It appears some of the files changed here now has validParams() declared twice with the addition of the forward declaration.

**Comment by andrsd** [comment]
@friedmud This will compile, will run and there will be no issues until `A` introduces its own params (ask me how I know ;-)):
```
class A : public MooseObject {};

class B : public A {};
---
registerObject(B);
---
validParams<B>() {
  return validParams<MooseObject>();
}
```

Of course, `validParams<B>()` is incorrect, but nothing can prevent people from doing this. So that's what I was asking. If this change will force users to have validParam for all classes and them catch the problem in `validParams<B>`.

**Comment by andrsd** [comment]
And I also have the same question as @gambka, why is forward declaration required? Having `validParams` after the class is valid C++.

File: framework/src/actions/MooseObjectAction.C (hunk 1 above)
Comment (andrsd) [reply framework/src/actions/MooseObjectAction.C]: Oh, the body is on lines 20-29. But why do we need this forward declaration here?

**Comment by friedmud** [comment]
@gambka @andrsd I just happened to pick a bad example 😛 I picked it because the filename was short... and since our convention (and the way it is in thousands of other header files) is to have it at the top... that's where I looked (and that's why we have the convention to put it at the top).

You are both correct that it's valid C++ to have it at the bottom but that is discouraged for multiple reasons (including this very reason).  In addition to just not being able to find/see it... it also makes sense to put it at the top because the definition of `validParams` in the `.C` is always at the top so it's nice to have parity between the header and source file.

Anyway - thanks for pointing out my mistake.  I fixed the 15 header files (out of ~100 I had to fix) that were done in this nonstandard way...

**Comment by friedmud** [comment]
I think this is good to go...

**Review by andrsd — APPROVED** [review]

**Comment by mangerij** [comment]
Just a reminder that Ferret is still a real thing. It yielded a publication in Nanoscale three days ago :)

Please just give me a heads up next time thanks! Or someone can put automatic notifications on the civet traffic. I try to keep up with PR's to try and understand what everyone is doing but I'm 8 hours ahead of you guys . Cheers

**Comment by permcody** [comment]
Sorry about that Jon, it may be time to add you to the "not allowed to fail" list! We can either provide you a patch or you can work on one. You just need to make sure your validParams are correct for all MooseObjects (i.e. at the top of all your files and declared in the header). That's all this PR does is enforce that.

**Comment by mangerij** [comment]
Yeah I'll fix it.. just thought we had a system. Guess its payback for how many times I broke CIVET in last few months 🥇 💯

**Comment by friedmud** [comment]
I did send out an email to both `moose-users` and `moose-announce`: https://groups.google.com/forum/#!topic/moose-announce/fp7NbELYGeU

We were just having a discussion about whether we should go submit a patch to you and I was thinking of doing it tomorrow.

To be 100% clear, all that's happening is that you're now actually able to see errors that have been in your code forever ;-)  This wasn't an API change or anything... just better error checking!

Let me put in a PR to you real quick...

**Comment by friedmud** [comment]
BTW: I _am_ actually doing this right now - so let's not duplicate effort!

**Comment by friedmud** [comment]
Ok Ferret PR is in...

**Comment by mangerij** [comment]
Thanks Derek!

**Pull request merged**